README.md
# sicinfo-gettext

### test

```# mocha test/gettext-test.js```

CHANGELOG.md
## sicinfo-gettext

[0.1.0] 2019-03-21 
- fork from ! gettext.js - Guillaume Potier - MIT Licensed - version 0.7


test/gettext-test.js
/**
 * application: myutils
 * 
 * powered by
 */
'use strict';

const 
  { expect } = require('chai'),
  i18n = require('../lib/gettext');

(function () {
  
    describe('gettext.js test suite', function () {
        describe('General API', function () {
            it('should be defined', function () {
                expect(i18n).to.be.a('function');
            });
            it('should be instanciable', function () {
                expect(i18n()).to.be.an('object');
            });
            it('should have default locale', function () {
                const _i18n = i18n();
                expect(_i18n.getLocale()).to.be.equal('en');
            });
            it('should allow to set locale', function () {
                const _i18n = i18n({
                    locale: 'fr'
                });
                expect(_i18n.getLocale()).to.be.equal('fr');
            });
            it('should allow to set messages', function () {
                const _i18n = i18n({
                    locale: 'fr',
                    messages: {
                        "apple": "pomme"
                    }
                });
                expect(_i18n.getLocale()).to.be.equal('fr');
                expect(_i18n.gettext('apple')).to.be.equal('pomme');
            });
        });
        describe('methods', function () {
            let _i18n;
            before(function () {
                _i18n = new i18n();
            });

            describe('strfmt', function () {
                it('should be a i18n method', function () {
                    expect(_i18n.strfmt).to.be.a('function');
                });
                it('should handle one replacement', function () {
                    expect(_i18n.strfmt('foo %1 baz', 'bar')).to.be.equal('foo bar baz');
                });
                it('should handle many replacements', function () {
                    expect(_i18n.strfmt('foo %1 baz %2', 'bar', 42)).to.be.equal('foo bar baz 42');
                });
                it('should handle order', function () {
                    expect(_i18n.strfmt('foo %2 baz %1', 'bar', 42)).to.be.equal('foo 42 baz bar');
                });
                it('should handle repeat', function () {
                    expect(_i18n.strfmt('foo %1 baz %1', 'bar', 42)).to.be.equal('foo bar baz bar');
                });
                it('should handle literal percent (%) signs', function () {
                    expect(_i18n.strfmt('foo 1%% bar')).to.be.equal('foo 1% bar');
                    expect(_i18n.strfmt('foo %1%% bar', 10)).to.be.equal('foo 10% bar');
                    expect(_i18n.strfmt('foo %%1 bar')).to.be.equal('foo %1 bar');
                    expect(_i18n.strfmt('foo %%%1 bar', 10)).to.be.equal('foo %10 bar');
                });
            });

            describe('expand_locale', function() {
                it('should be a i18n method', function() {
                    expect(_i18n.expand_locale).to.be.a('function');
                });
                it('should handle simple locale', function() {
                    expect(_i18n.expand_locale('fr')).to.eql(['fr']);
                });
                it('should handle complex locale', function() {
                    expect(_i18n.expand_locale('de-CH-1996')).to.eql(['de-CH-1996', 'de-CH', 'de']);
                });
            });

            describe('gettext', function () {
                it('should handle peacefully singular untranslated keys', function () {
                    expect(_i18n.gettext('not translated')).to.be.equal('not translated');
                });
                it('should handle peacefully singular untranslated keys with extra', function () {
                    expect(_i18n.gettext('not %1 translated', 'correctly')).to.be.equal('not correctly translated');
                    expect(_i18n.gettext('not %1 %2 translated', 'fully', 'correctly')).to.be.equal('not fully correctly translated');
                });
                it('should fallback to father language', function() {
                    _i18n = new i18n();

                    _i18n.setMessages('messages', 'fr', {
                        "Mop": "Serpillière",
                    });

                    _i18n.setMessages('messages', 'fr-BE', {
                        "Mop": "Torchon",
                    });

                    _i18n.setLocale('fr-BE');
                    expect(_i18n.gettext("Mop")).to.be.equal("Torchon");

                    _i18n.setLocale('fr');
                    expect(_i18n.gettext("Mop")).to.be.equal("Serpillière");

                    _i18n.setLocale('fr-FR');
                    expect(_i18n.gettext("Mop")).to.be.equal("Serpillière");
                });
            });

            describe('ngettext', function () {
                it('should handle peacefully plural untranslated keys', function () {
                    // english default plural rule is n !== 1
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 0)).to.be.equal('0 not translated plural');
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 1)).to.be.equal('1 not translated singular');
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 3)).to.be.equal('3 not translated plural');
                });
                it('should handle peacefully plural untranslated keys with extra', function () {
                    expect(_i18n.ngettext('%1 not %2 singular', '%1 not %2 plural', 1, 'foo')).to.be.equal('1 not foo singular');
                    expect(_i18n.ngettext('%1 not %2 singular', '%1 not %2 plural', 3, 'foo')).to.be.equal('3 not foo plural');
                });
                it('should use default english plural form for untranslated keys', function () {
                    _i18n = new i18n({ locale: 'fr', plural_forms: 'nplurals=2; plural=n>1;' });
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 0)).to.be.equal('0 not translated plural');
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 1)).to.be.equal('1 not translated singular');
                    expect(_i18n.ngettext('%1 not translated singular', '%1 not translated plural', 3)).to.be.equal('3 not translated plural');
                });
                it('should handle correctly other language plural passed through setMessages method', function () {
                    _i18n = new i18n({locale: 'fr'});
                    _i18n.setMessages('messages', 'fr', {
                        "There is %1 apple": [
                            "Il y a %1 pomme",
                            "Il y a %1 pommes"
                        ]
                    }, 'nplurals=2; plural=n>1;');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 0)).to.be.equal('Il y a 0 pomme');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 1)).to.be.equal('Il y a 1 pomme');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 2)).to.be.equal('Il y a 2 pommes');
                });
                it('should handle correctly other language plural passed through init options', function () {
                    _i18n = new i18n({
                        locale: 'fr',
                        messages: {
                            "There is %1 apple": [
                                "Il y a %1 pomme",
                                "Il y a %1 pommes"
                            ]
                        },
                        plural_forms: 'nplurals=2; plural=n>1;'
                    });
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 0)).to.be.equal('Il y a 0 pomme');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 1)).to.be.equal('Il y a 1 pomme');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 2)).to.be.equal('Il y a 2 pommes');
                });
                it('should ignore a plural translation when requesting the singular form', function () {
                    _i18n = new i18n({ locale: 'fr' });
                    _i18n.setMessages('messages', 'fr', {
                        "apple": [
                            "pomme",
                            "pommes"
                        ]
                    }, 'nplurals=2; plural=n>1;');
                    expect(_i18n.gettext('apple')).to.be.equal('apple');
                    expect(_i18n.ngettext('apple', 'apples', 1)).to.be.equal('pomme');
                    expect(_i18n.ngettext('apple', 'apples', 2)).to.be.equal('pommes');
                });
                it('should ignore a singular translation when requesting the plural form', function () {
                    _i18n = new i18n({ locale: 'fr' });
                    _i18n.setMessages('messages', 'fr', {
                        "apple": "pomme"
                    });
                    expect(_i18n.gettext('apple')).to.be.equal('pomme');
                    expect(_i18n.ngettext('apple', 'apples', 1)).to.be.equal('apple');
                    expect(_i18n.ngettext('apple', 'apples', 2)).to.be.equal('apples');
                });
                it('should fail unvalid plural form', function () {
                    _i18n = new i18n({ locale: 'foo' });
                    _i18n.setMessages('messages', 'foo', {
                        "There is %1 apple": [
                            "Il y a %1 pomme",
                            "Il y a %1 pommes"
                        ]
                    }, 'nplurals=2; plural=[not valid];');

                    // do not throw error on default plural form if key does not have a translation
                    expect(_i18n.ngettext('foo', 'bar', 2)).to.be.equal('bar');

                    try {
                        _i18n.ngettext('There is %1 apple', 'There are %1 apples', 42);
                    } catch (e) {
                        expect(e.message).to.be.equal('The plural form "nplurals=2; plural=[not valid];" is not valid');
                    }
                });
                it('should handle multiple locale & pluals cohabitation', function () {
                    _i18n = new i18n({ locale: 'foo' });
                    _i18n.setMessages('messages', 'foo', {
                        "singular": [
                            "singular",
                            "plural"
                        ]
                    }, 'nplurals=2; plural=n>10;');
                    _i18n.setMessages('messages', 'bar', {
                        "singular": [
                            "singulier",
                            "pluriel"
                        ]
                    }, 'nplurals=2; plural=n>100;');
                    expect(_i18n.ngettext('singular', 'plural', 9)).to.be.equal('singular');
                    expect(_i18n.ngettext('singular', 'plural', 11)).to.be.equal('plural');

                    _i18n.setLocale('bar');
                    expect(_i18n.ngettext('singular', 'plural', 9)).to.be.equal('singulier');
                    expect(_i18n.ngettext('singular', 'plural', 11)).to.be.equal('singulier');
                    expect(_i18n.ngettext('singular', 'plural', 111)).to.be.equal('pluriel');
                });
                it('should fallback to singular form if there is a problem with plurals', function () {
                    // incorrect plural, more than nplurals
                    _i18n = new i18n({ locale: 'foo' });
                    _i18n.setMessages('messages', 'foo', {
                        "apple": [
                            "pomme",
                            "pommes"
                        ]
                    }, 'nplurals=2; plural=3;');
                    expect(_i18n.ngettext('apple', 'apples', 1)).to.be.equal('pomme');

                    // plural is correct, but according to nplurals there should be more translations
                    _i18n = new i18n({ locale: 'ru' });
                    _i18n.setMessages('messages', 'ru', {
                        "%1 apple": [
                            "%1 яблоко",
                            "%1 яблока"
                            // "%1 яблок" - missed translation
                        ]
                    }, "nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2);");
                    expect(_i18n.ngettext('%1 apple', '%1 apples', 5)).to.be.equal('5 яблоко');
                });
            });

            describe('loadJSON', function () {
                it('should properly load json', function () {
                    var parsedJSON = {
                        "": {
                            "language": "fr",
                            "plural-forms": "nplurals=2; plural=n>1;"
                        },
                        "Loading...": "Chargement...",
                        "Save": "Sauvegarder",
                        "There is %1 apple": [
                            "Il y a %1 pomme",
                            "Il y a %1 pommes"
                        ],
                        "Checkout\u0004Save": "Sauvegarder votre panier"
                    },
                    _i18n = i18n()
                        .loadJSON(JSON.stringify(parsedJSON))
                        .setLocale('fr');
                    expect(_i18n.getLocale(), 'fr');
                    expect(_i18n.textdomain(), 'messages');
                    expect(_i18n.gettext('Save')).to.be.equal('Sauvegarder');
                    expect(_i18n.gettext('Loading...')).to.be.equal('Chargement...');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 0)).to.be.equal('Il y a 0 pomme');
                    expect(_i18n.ngettext('There is %1 apple', 'There are %1 apples', 2)).to.be.equal('Il y a 2 pommes');
                });
            });
        });
    });
}());

lib/gettext.js
/**
 * application: gettext
 * 
 * powered by ! gettext.js - Guillaume Potier - MIT Licensed - version 0.7
 * 
 * updated by Moreira in 2019-03-21
 */
'use strict';

(function (root, undef) {
  
  const 
  
  _plural_funcs = {},
  _dictionary = {},
  _plural_forms = {},
  
  
  defaults = {
    domain: 'messages',
    locale: (typeof document !== 'undefined' && document.documentElement.getAttribute('lang')) || 'en',
    plural_func: n => ({ nplurals: 2, plural: (n!=1) ? 1 : 0 }),
    ctxt_delimiter: String.fromCharCode(4) // \u0004
  },

  
  // sprintf equivalent, takes a string and some arguments to make a computed string
  // eg: strfmt("%1 dogs are in %2", 7, "the kitchen"); => "7 dogs are in the kitchen"
  // eg: strfmt("I like %1, bananas and %1", "apples"); => "I like apples, bananas and apples"
  strfmt = (...args) => {
    return args[0]
      // put space after double % to prevent placeholder replacement of such matches
      .replace(/%%/g, '%% ')
      // replace placeholders
      .replace(/%(\d+)/g, (str, p1) => args[p1])
      // replace double % and space with single %
      .replace(/%% /g, '%');
  },
      

  // handy mixins taken from underscode.js
  isObject = obj => {
    const type = typeof obj;
    return type === 'function' || type === 'object' && !!obj;
  },

  
  expand_locale = locale => {
    const locales = [locale];
    let i = locale.lastIndexOf('-');
    while (i > 0) {
      locale = locale.slice(0, i);
      locales.push(locale);
      i = locale.lastIndexOf('-');
    }
    return locales;
  },
  
  
  getPluralFunc = plural_form => {
    
    // Plural form string regexp
    // taken from https://github.com/Orange-OpenSource/gettext.js/blob/master/lib.gettext.js
    // plural forms list available here http://localization-guide.readthedocs.org/en/latest/l10n/pluralforms.html
    const pf_re = new RegExp('^\\s*nplurals\\s*=\\s*[0-9]+\\s*;\\s*plural\\s*=\\s*(?:\\s|[-\\?\\|&=!<>+*/%:;n0-9_\(\)])+');
    if (!pf_re.test(plural_form)) throw new Error(strfmt('The plural form "%1" is not valid', plural_form));
    
    // Careful here, this is a hidden eval() equivalent..
    // Risk should be reasonable though since we test the plural_form through regex before
    // taken from https://github.com/Orange-OpenSource/gettext.js/blob/master/lib.gettext.js
    // TODO: should test if https://github.com/soney/jsep present and use it if so
    return new Function("n", `
      var plural, nplurals; 
      ${plural_form} return { 
        nplurals: nplurals, 
        plural: (plural === true ? 1 : (plural ? plural : 0)) 
      };`
    );
  },
  
  
  // Proper translation function that handle plurals and directives
  // Contains juicy parts of https://github.com/Orange-OpenSource/gettext.js/blob/master/lib.gettext.js
  t = function (_locale, messages, n, options, ...args) {
    
    // Singular is very easy, just pass dictionnary message through strfmt
    if (1 === messages.length)
      return strfmt(...[messages[0]].concat(args));
      
    let plural;
    
    // if a plural func is given, use that one
    if (options.plural_func) {
      plural = options.plural_func(n);
      
      // if plural form never interpreted before, do it now and store it
    } else if (!_plural_funcs[_locale]) {

      _plural_funcs[_locale] = getPluralFunc(_plural_forms[_locale]);
      plural = _plural_funcs[_locale](n);

    // we have the plural function, compute the plural result
    } else {
      plural = _plural_funcs[_locale](n);
    }
    
    // If there is a problem with plurals, fallback to singular one
    if ('undefined' === typeof plural.plural || plural.plural > plural.nplurals || messages.length <= plural.plural) 
      plural.plural = 0;

    return strfmt(...[messages[plural.plural], n].concat(args));
  },
  
  
  setMessages = function (domain, locale, messages, plural_forms) {
    if (!domain || !locale || !messages)
      throw new Error('You must provide a domain, a locale and messages');
      
    if ('string' !== typeof domain || 'string' !== typeof locale || !isObject(messages))
      throw new Error('Invalid arguments');
      
    if (plural_forms) _plural_forms[locale] = plural_forms;
    if (!_dictionary[domain]) _dictionary[domain] = {};
    _dictionary[domain][locale] = messages;

    return this;
  },
  
  
  loadJSON = function (jsonData, domain) {
    if (!isObject(jsonData))
      jsonData = JSON.parse(jsonData);

    if (!jsonData[''] || !jsonData['']['language'] || !jsonData['']['plural-forms'])
      throw new Error('Wrong JSON, it must have an empty key ("") with "language" and "plural-forms" information');

    const headers = jsonData[''];
    delete jsonData[''];

    return setMessages.call(this, domain || defaults.domain, headers['language'], jsonData, headers['plural-forms']);
  },


  i18n = function (options = {}) {
    // this.__version = '0.5.3';

    // default values that could be overriden in i18n() construct
    let 
      _locale = options.locale || defaults.locale,
      _domain = options.domain || defaults.domain,
      _ctxt_delimiter = options.ctxt_delimiter || defaults.ctxt_delimiter;

    if (options.messages) {
      _dictionary[_domain] = {};
      _dictionary[_domain][_locale] = options.messages;
    }

    if (options.plural_forms) {
      _plural_forms[_locale] = options.plural_forms;
    }

    return {
      strfmt: strfmt, // expose strfmt util
      expand_locale: expand_locale, // expose expand_locale util

      // Declare shortcuts
      __: function (...args) { return this.gettext(...args) },
      _n: function (...args) { return this.ngettext(...args) },
      _p: function (...args) { return this.pgettext(...args) },

      setMessages: setMessages,
      loadJSON: loadJSON,
      setLocale: function (locale) {
        _locale = locale;
        return this;
      },
      getLocale: function () {
        return _locale;
      },
      // getter/setter for domain
      textdomain: function (domain) {
        if (!domain) return _domain;
        _domain = domain;
        return this;
      },
      gettext: function (msgid, ...args) {
        return this.dcnpgettext(...[undef, undef, msgid, undef, undef].concat(args));
      },
      ngettext: function (msgid, msgid_plural, n, ...args) {
        return this.dcnpgettext(...[undef, undef, msgid, msgid_plural, n].concat(args));
      },
      pgettext: function (msgctxt, msgid, ...args) {
        return this.dcnpgettext.apply(this, [undef, msgctxt, msgid, undef, undef].concat(args));
      },
      dcnpgettext: function (domain, msgctxt, msgid, msgid_plural, n, ...args) {
        
        if (domain == undef) domain = _domain;

        if ('string' !== typeof msgid)
          throw new Error(this.strfmt('Msgid "%1" is not a valid translatable string', msgid));

        let
          translation,
          options = {},
          key = msgctxt ? `${msgctxt}${_ctxt_delimiter}${msgid}` : msgid,
          exist,
          locale,
          locales = expand_locale(_locale);
        
        for (var i in locales) {
          locale = locales[i];
          exist = _dictionary[domain] && _dictionary[domain][locale] && _dictionary[domain][locale][key];

          // because it's not possible to define both a singular and a plural form of the same msgid,
          // we need to check that the stored form is the same as the expected one.
          // if not, we'll just ignore the translation and consider it as not translated.
          if (msgid_plural) {
            exist = exist && "string" !== typeof _dictionary[domain][locale][key];
          } else {
            exist = exist && "string" === typeof _dictionary[domain][locale][key];
          }
            
          if (exist) break;
        }

        if (!exist) {
          translation = msgid;
          options.plural_func = defaults.plural_func;
        } else {
          translation = _dictionary[domain][locale][key];
        }

        return msgid_plural ?
          
          // Plural one
          t(...[_locale, exist ? translation : [msgid, msgid_plural], n, options].concat(args)) :
          
          // Singular form
          t(...[_locale, [translation], n, options].concat(args));

      }
    };
  };

  // Handle node, commonjs
  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports)
      exports = module.exports = i18n;
    exports.i18n = i18n;

  // Handle AMD
  } else if (typeof define === 'function' && define.amd) {
    define(function() { return i18n; });

  // Standard window browser
  } else {
    root['i18n'] = i18n;
  }
  
})(this);